onnxruntime/core/providers/webnn/builders/impl/dropout_op_builder.cc

```diff
@@ -0,0 +1,100 @@
+
+namespace onnxruntime {
+namespace webnn {
+
+class DropoutOpBuilder : public BaseOpBuilder {
+  // Add operator related.
+ public:
+  void AddInitializersToSkip(ModelBuilder& model_builder, const Node& node) const override;
+
+ private:
+  Status AddToModelBuilderImpl(ModelBuilder& model_builder, const Node& node,
+                               const logging::Logger& logger) const override ORT_MUST_USE_RESULT;
+
+  // Operator support related.
+ private:
+  bool IsOpSupportedImpl(const InitializedTensorSet& initializers, const Node& node,
+                         const WebnnDeviceType /* device_type */, const logging::Logger& logger) const override;
+};
+
+// Add operator related.
+
+void DropoutOpBuilder::AddInitializersToSkip(ModelBuilder& model_builder, const Node& node) const {
+  // Skip ratio and training_mode if present.
+  for (size_t i = 1; i < node.InputDefs().size(); i++) {
+    model_builder.AddInitializerToSkip(node.InputDefs()[i]->Name());
+    model_builder.AddInputToSkip(node.InputDefs()[i]->Name());
+  }
+}
+
+Status DropoutOpBuilder::AddToModelBuilderImpl(ModelBuilder& model_builder,
+                                               const Node& node,
+                                               const logging::Logger& logger) const {
+  const auto& input_defs = node.InputDefs();
+  const auto& output_defs = node.OutputDefs();
+  emscripten::val input = model_builder.GetOperand(input_defs[0]->Name());
+  emscripten::val options = emscripten::val::object();
+  options.set("label", node.Name());
+
+  // WebNN EP only supports test mode, we don't need to care about other inputs or
+  // attributes about training mode. Simply use WebNN's identity op to copy the input.
+  emscripten::val output = model_builder.GetBuilder().call<emscripten::val>("identity", input, options);
+
+  model_builder.AddOperand(output_defs[0]->Name(), std::move(output));
+
+  // If mask output is requested as output it will contain all ones (bool tensor).
+  if (output_defs.size() > 1) {
+    std::vector<int64_t> mask_shape;
+    ORT_RETURN_IF_NOT(GetShape(*output_defs[1], mask_shape, logger), "Cannot get mask output's shape");
+    std::vector<uint32_t> dims = GetVecUint32FromVecInt64(mask_shape);
+
+    emscripten::val desc = emscripten::val::object();
+    desc.set("dataType", "uint8");
+    desc.set("dimensions", emscripten::val::array(dims));
+    const auto num_elements = narrow<uint32_t>(Product(mask_shape));
+    emscripten::val ones_buffer = emscripten::val::global("Uint8Array").new_(num_elements);
+    ones_buffer.call<void>("fill", 1);
+
+    emscripten::val mask_output = model_builder.GetBuilder().call<emscripten::val>("constant", desc, ones_buffer);
+
+    emscripten::val options = emscripten::val::object();
+    options.set("label", output_defs[1]->Name() + "_identity");
+    // Add additional identity op in case the mask is the output of a WebNN graph,
+    // beacuse WebNN does not support a constant operand as output.
```

Comment:
Would you consider this a spec issue that we should raise?

Reply:
I don't have a strong opinion, we don't have a real use case that provides a known constant as output.

Reply:
Well, even little implementation woes like this are legitimate considerations for callers (it doesn't necessarily have to be in the *original model graph* for it to be considered a real use case, as even by products of graph processing matter). Coincidentally a coworker just yesterday came to my office to tell me about a bug where ORT was passing a subgraph downward, and thanks to an identity elimination optimization along the way, it was yielding an empty graph and causing an error with a component that was not prepared for empty graphs (which ideally should have just treated it as a nop :/). Note this was actually in the DML EP, but I foresee the same issue in WebNN for callers. Now, I can't find the WebNN issue before where we discussed the legitimacy of empty graphs before (maybe @huningxin does), but at the time we decided to require the caller to insert an identity explicitly, like you did here, which isn't that hard, but it's still an extra potential gotcha that API callers need to know about.